### PR TITLE
Extract logic to lib/metrics.go and create `cmd/prom2json` directory

### PR DESCRIPTION
**Pull request by ChristianKniep**
Rebased PR of #10

### CODE REVIEW

**Comment by ChristianKniep** [comment]
Not sure if the `vendor.json` is ok like that, as I am refering `ChristianKniep/prom2json`. Otherwise I hope I fetch all the comments in #10.

File: vendor/vendor.json [hunk 1]
```diff
@@ -45,17 +45,23 @@
...
+		},
+		{
+			"checksumSHA1": "To5eWGIjHQcw3pQ0N1uFepqaxmE=",
+			"path": "golang.org/x/sys/unix",
+			"revision": "739734461d1c916b6c72a63d7efda2b27edb369f",
+			"revisionTime": "2017-07-07T11:51:12Z"
 		}
 	],
-	"rootPath": "github.com/prometheus/prom2json"
+	"rootPath": "github.com/ChristianKniep/prom2json"
```
Comment (beorn7) [inline vendor/vendor.json]: That's an artifact from the fork. I'm afraid you have to manually set this back to `"github.com/prometheus/prom2json"`.

File: cmd/prom2json/main_test.go [hunk 2]
```diff
@@ -7,21 +7,22 @@ import (
 
 	"github.com/davecgh/go-spew/spew"
 	dto "github.com/prometheus/client_model/go"
+	p2j "github.com/ChristianKniep/prom2json"
```
Comment (beorn7) [inline cmd/prom2json/main_test.go]: This has to be `"github.com/prometheus/prom2json"`. Sorry for the hassle wit the fork.

Comment (beorn7) [inline cmd/prom2json/main_test.go]: Also, please have a blank line in front. (We separate imports from the same repo.)

File: cmd/prom2json/main.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	"runtime"
+
+	"github.com/prometheus/log"
+
+	dto "github.com/prometheus/client_model/go"
+	"github.com/ChristianKniep/prom2json"
+)
+
+func main() {
+	runtime.GOMAXPROCS(2)
```
Comment (beorn7) [inline cmd/prom2json/main.go]: This was 5 before. Why change it?

In different news, with modern Go versions, we don't even need this anymore. I'd just remove the line.

**Comment by beorn7** [comment]
Is there any special reason why you put the tests as `cmd/prom2json/main_test.go`? I'd expect them to be next to the library code they test.

File: vendor/vendor.json (hunk 1 above)
Comment (ChristianKniep) [inline vendor/vendor.json]: true, as I moved the test to the root, the import vanishes anyway.

**Comment by ChristianKniep** [comment]
Moved the test to the root and renamed the imports (plus blank). Thanks @beorn7 for you endurance. :)

**Comment by beorn7** [comment]
Thanks for your patience. I think this looks good over-all. I have only nits about doc comments. Will post the comments soon.

File: vendor/vendor.json (hunk 1 above)
Comment (beorn7) [inline vendor/vendor.json]: But this entry is still here.

**Comment by ChristianKniep** [comment]
Cool, should I rename `metric.go` to `prom2json.go` or is it fine like that...?

File: metric.go [hunk 4]
```diff
@@ -1,89 +1,78 @@
...
+	"github.com/prometheus/log"
 
 	dto "github.com/prometheus/client_model/go"
+
 )
 
 const acceptHeader = `application/vnd.google.protobuf;proto=io.prometheus.client.MetricFamily;encoding=delimited;q=0.7,text/plain;version=0.0.4;q=0.3`
 
-type metricFamily struct {
+// Family holds the s group of metrics/summary/histogram of a certain name...
```
Comment (beorn7) [inline metric.go]: `s`?

I think doc comments here don't have to explain concepts from the Prometheus data model.

How about:
`// Family mirrors the MetricFamily proto message.`

File: metric.go [hunk 5]
```diff
@@ -1,89 +1,78 @@
...
+	//Time    time.Time
 	Name    string        `json:"name"`
 	Help    string        `json:"help"`
 	Type    string        `json:"type"`
 	Metrics []interface{} `json:"metrics,omitempty"` // Either metric or summary.
 }
 
-// metric is for all "single value" metrics.
-type metric struct {
+// Metric is for all "single value" metrics.
```
Comment (beorn7) [inline metric.go]: Add: `, i.e. Counter, Gauge, and Untyped.`

File: metric.go [hunk 6]
```diff
@@ -1,89 +1,78 @@
...
+type Metric struct {
 	Labels map[string]string `json:"labels,omitempty"`
 	Value  string            `json:"value"`
 }
 
-type summary struct {
+// Summary describes a summary metric
```
Comment (beorn7) [inline metric.go]: `// Summary mirrors the Summary proto message.`

(Note that sentences should end with a period.)

File: metric.go [hunk 7]
```diff
@@ -1,89 +1,78 @@
...
+type Summary struct {
 	Labels    map[string]string `json:"labels,omitempty"`
 	Quantiles map[string]string `json:"quantiles,omitempty"`
 	Count     string            `json:"count"`
 	Sum       string            `json:"sum"`
 }
 
-type histogram struct {
+// Histogram holds buckets of metrics
```
Comment (beorn7) [inline metric.go]: `// Histogram mirrors the Histogram proto message.`

File: metric.go [hunk 8]
```diff
@@ -1,89 +1,78 @@
...
+type Histogram struct {
 	Labels  map[string]string `json:"labels,omitempty"`
 	Buckets map[string]string `json:"buckets,omitempty"`
 	Count   string            `json:"count"`
 	Sum     string            `json:"sum"`
 }
 
-func newMetricFamily(dtoMF *dto.MetricFamily) *metricFamily {
-	mf := &metricFamily{
+// NewFamily consumes a MetricFamily and transforms it to the local Family type
```
Comment (beorn7) [inline metric.go]: End with period.

File: metric.go [hunk 9]
```diff
@@ -129,40 +118,27 @@ func makeBuckets(m *dto.Metric) map[string]string {
 	return result
 }
 
-func fetchMetricFamilies(url string, ch chan<- *dto.MetricFamily, certificate string, key string) {
+// FetchMetricFamilies does stuff
```
Comment (beorn7) [inline metric.go]: `// FetchMetricFamilies retrieves metrics from the provided URL, decodes them into MetricFamily proto messages, and sends them to the provided channel. It returns after all MetricFamilies have been sent.`

File: metric.go [hunk 10]
```diff
@@ -129,40 +118,27 @@ func makeBuckets(m *dto.Metric) map[string]string {
...
 		log.Fatalf("executing GET request for URL %q failed: %s", url, err)
 	}
 	defer resp.Body.Close()
 	if resp.StatusCode != http.StatusOK {
 		log.Fatalf("GET request for URL %q returned HTTP status %s", url, resp.Status)
 	}
+	ParseResponse(resp, ch)
+}
+
+// ParseResponse consumes an http.Response and pushes it to the MetricFamily channel
```
Comment (beorn7) [inline metric.go]: End with period. Also, add the return behavior (see above, returns when all MetricFamilies have been sent to the channel).

File: metric.go [hunk 11]
```diff
@@ -192,29 +168,14 @@ func decodeContent(client *http.Client, url string, ch chan<- *dto.MetricFamily)
...
-	for mf := range mfChan {
-		result = append(result, newMetricFamily(mf))
-	}
-	json, err := json.Marshal(result)
-	if err != nil {
-		log.Fatalln("error marshaling JSON:", err)
-	}
-	if _, err := os.Stdout.Write(json); err != nil {
-		log.Fatalln("error writing to stdout:", err)
+// AddLabel allows to add key/value labels to an already existing Family
```
Comment (beorn7) [inline metric.go]: End with period.

**Comment by beorn7** [comment]
> Cool, should I rename metric.go to prom2json.go or is it fine like that...?

I don't feel strongly, but perhaps that's better.

Since the package is not called `metric`, perhaps also keep `MetricFamily` as a name rather than just `Family`? But again, I don't feel strongly.

**Comment by ChristianKniep** [comment]
Ok, I'll keep the naming for the time being.

**Comment by beorn7** [comment]
Cool, thanks.

**Pull request merged**